### PR TITLE
Fixed some follow ups after restructuring and issues around generator-theia

### PR DESCRIPTION
**Pull request by akosyakov**

### CODE REVIEW

File: doc/Developing.md [hunk 1]
```diff
@@ -41,7 +41,7 @@ This command does a few things:
 
 ## Run the browser-based example application
 
-We can start the application with:
+We can start the application from examples/browser directory with:
```
Comment (simark) [inline doc/Developing.md]: I'd suggest saying "from the".

File: doc/Developing.md [hunk 2]
```diff
@@ -50,7 +50,16 @@ tab with the frontend.
...
+From the root directory run:
+
+    npm run rebuild:electron
+
+This command rebuilds native node packages against the version of node used by electron.
+To rollback changes run:
+
+    npm run rebuild:web
+
+It can also be started with from examples/electron directory:
```
Comment (simark) [inline doc/Developing.md]: I'd change it to

    It can also be started from the examples/electron directory with:

File: generator-theia/src/common/abstract-generator.ts [hunk 3]
```diff
@@ -51,7 +51,7 @@ export abstract class AbstractGenerator {
                 return `Promise.resolve(${invocation})`;
             }
             return invocation;
-        }).map(statement => `.then(function () { return ${statement}.then(load) })`);
+        }).map(statement => `    .then(function () { return ${statement}.then(load) })`);
```
Comment (simark) [inline generator-theia/src/common/abstract-generator.ts]: What is this change for?

Comment (akosyakov) [inline generator-theia/src/common/abstract-generator.ts]: to fix indentation in the generated source files: `src-gen/frontend/main.js` and `src-gen/backend/main.js`

Comment (hexa00) [inline generator-theia/src/common/abstract-generator.ts]: Yes thx for that! I had run into trouble with it

File: .travis.yml [hunk 4]
```diff
@@ -21,6 +21,7 @@ before_script:
 before_install:
   - "npm install -g npm@^4"
   - "google-chrome-stable --headless --disable-gpu --remote-debugging-port=9222 http://localhost &"
+bundler_args: --retry 0
```
Comment (hexa00) [inline .travis.yml]: Is this travis specific or it's passed to lerna or something ? 
I can't find options on lerna or npm and travis ci references bundler_args with the ruby bundler

File: doc/Developing.md (hunk 2 above)
Comment (hexa00) [inline doc/Developing.md]: I don't see why someone would want to rollback however ? 
Is this more confusing then helping?

Maybe move this to the web basec example section?

File: doc/Developing.md [hunk 5]
```diff
@@ -116,16 +125,17 @@ To build and run the browser example:
 
     git clone https://github.com/theia-ide/theia \
     && cd theia \
-    && npm install --unsafe-perm\
-    && cd ../../examples/browser \
+    && npm install --unsafe-perm \
```
Comment (hexa00) [inline doc/Developing.md]: BTW we discussed this a bit and --unsafe-perm looks really weird.
I Suggest we move it out of the TL;DR and add it to the troubleshooting section
but that can be in another PR

File: doc/Internals.md [hunk 6]
```diff
@@ -198,20 +198,6 @@ And it returns the loggerServer as the object that will be exposed over JSON-RPC
...
-
-``` javascript
-            '/logger/*': {
-                target: 'ws://localhost:3000',
-                ws: true
-            },
-```
-
-To webpack.config.js so that the requests are proxied to the backend properly.
-
```
Comment (hexa00) [inline doc/Internals.md]: We should update it to say that now the endpoints with /services/* are served by the webpack dev server

File: examples/browser/webpack.config.js [hunk 7]
```diff
@@ -64,18 +63,13 @@ module.exports = {
     resolve: {
         extensions: ['.js'],
         alias: {
-            'vs': path.resolve(outputPath, monacoEditorPath),
-            'dtrace-provider': emptyPath,
-            'safe-json-stringify': emptyPath,
-            'mv': emptyPath,
-            'source-map-support': emptyPath
+            'vs': path.resolve(outputPath, monacoEditorPath)
```
Comment (hexa00) [inline examples/browser/webpack.config.js]: Why is this not needed for bunyan anymore?

File: package.json [hunk 8]
```diff
@@ -10,11 +13,13 @@
     "preinstall": "cd generator-theia && npm install",
     "prepare": "npm install generator-theia && npm run bootstrap && npm run clean && npm run build",
     "prepublishOnly": "npm run test",
-    "bootstrap": "node scripts/generate.js && lerna bootstrap --hoist",
+    "bootstrap": "node scripts/generate.js && lerna bootstrap --hoist -- --no-optional",
```
Comment (hexa00) [inline package.json]: What is this doing? (the --no-optional) I'm not sure it's passed to what

File: examples/browser/webpack.config.js (hunk 7 above)
Comment (akosyakov) [inline examples/browser/webpack.config.js]: We don't use bunyan in the frontend and use the webpack to build only the frontend, so these modules are never bundled

File: package.json (hunk 8 above)
Comment (akosyakov) [inline package.json]: Bunyan has some optional dependencies for dtracing and so on, we don't use it, by passing `--no-optional` we avoid downloading unnecessary dependencies.

File: .travis.yml (hunk 4 above)
Comment (akosyakov) [inline .travis.yml]: It is travis specific, depending on provided language travis knows how to build the project, like run `npm install` first and then `npm test` for node projects. It also has some defaults for such projects, e.g. for node to try to run `npm install` 3 times if the first fails, since npmjs sometimes is not available.

File: doc/Developing.md (hunk 5 above)
Comment (akosyakov) [inline doc/Developing.md]: Could you open an issue for it?

File: doc/Developing.md (hunk 2 above)
Comment (akosyakov) [inline doc/Developing.md]: After rebuilding for electron, web won't work, but by default web is working. I can put in the web section that if you run `npm run rebuild:electron` don't forget to run `npm run rebuild:web` fisrt.

Comment (hexa00) [inline doc/Developing.md]: OK yes I was curious about that so this means that since with hoisting we have only 1 node-modules we can only have 1 example at the time without rebuilding :(
But yes sounds better to me in the web section thx

File: .travis.yml (hunk 4 above)
Comment (hexa00) [inline .travis.yml]: OK thanks!

File: doc/Developing.md (hunk 5 above)
Comment (hexa00) [inline doc/Developing.md]: Issue #311

File: examples/browser/webpack.config.js (hunk 7 above)
Comment (hexa00) [inline examples/browser/webpack.config.js]: Right :) thx

File: doc/Developing.md (hunk 2 above)
Comment (akosyakov) [inline doc/Developing.md]: It could be that we can have a setup there we exclude node-pty from hoisting and don't need to run rebuild, but I've spent half a day and failed to come up with such. I think mostly we work with one or another example anyway and later with more experience maybe can overcome it.

File: package.json (hunk 8 above)
Comment (hexa00) [inline package.json]: OK I still don't see how it goes up to bunyan but will investigate

Comment (akosyakov) [inline package.json]: https://github.com/trentm/node-bunyan/blob/master/package.json#L25 it declares optional dependencies here

Comment (hexa00) [inline package.json]: So this is passed to npm install in the end? so all modules are with --no-optional?

File: doc/Developing.md (hunk 2 above)
Comment (hexa00) [inline doc/Developing.md]: Sounds good

File: package.json (hunk 8 above)
Comment (akosyakov) [inline package.json]: right, to enable them later we can either remove passing this option or add them explicitly
